<html>

<head>
<meta http-equiv=Content-Type content="text/html; charset=windows-1252">
<meta name=Generator content="Microsoft Word 15 (filtered)">
<style>
<!--
 /* Font Definitions */
 @font-face
	{font-family:"Cambria Math";
	panose-1:2 4 5 3 5 4 6 3 2 4;}
@font-face
	{font-family:Calibri;
	panose-1:2 15 5 2 2 2 4 3 2 4;}
@font-face
	{font-family:"Calibri Light";
	panose-1:2 15 3 2 2 2 4 3 2 4;}
 /* Style Definitions */
 p.MsoNormal, li.MsoNormal, div.MsoNormal
	{margin-top:0cm;
	margin-right:0cm;
	margin-bottom:10.0pt;
	margin-left:0cm;
	line-height:115%;
	font-size:11.0pt;
	font-family:"Calibri",sans-serif;}
h2
	{mso-style-link:"Heading 2 Char";
	margin-top:10.0pt;
	margin-right:0cm;
	margin-bottom:0cm;
	margin-left:0cm;
	margin-bottom:.0001pt;
	line-height:115%;
	page-break-after:avoid;
	font-size:13.0pt;
	font-family:"Calibri Light",sans-serif;
	color:#4472C4;
	font-weight:bold;}
p.MsoCaption, li.MsoCaption, div.MsoCaption
	{margin-top:0cm;
	margin-right:0cm;
	margin-bottom:10.0pt;
	margin-left:0cm;
	font-size:9.0pt;
	font-family:"Calibri",sans-serif;
	color:#4472C4;
	font-weight:bold;}
p
	{margin-right:0cm;
	margin-left:0cm;
	font-size:12.0pt;
	font-family:"Times New Roman",serif;}
p.MsoListParagraph, li.MsoListParagraph, div.MsoListParagraph
	{margin-top:0cm;
	margin-right:0cm;
	margin-bottom:10.0pt;
	margin-left:36.0pt;
	line-height:115%;
	font-size:11.0pt;
	font-family:"Calibri",sans-serif;}
p.MsoListParagraphCxSpFirst, li.MsoListParagraphCxSpFirst, div.MsoListParagraphCxSpFirst
	{margin-top:0cm;
	margin-right:0cm;
	margin-bottom:0cm;
	margin-left:36.0pt;
	margin-bottom:.0001pt;
	line-height:115%;
	font-size:11.0pt;
	font-family:"Calibri",sans-serif;}
p.MsoListParagraphCxSpMiddle, li.MsoListParagraphCxSpMiddle, div.MsoListParagraphCxSpMiddle
	{margin-top:0cm;
	margin-right:0cm;
	margin-bottom:0cm;
	margin-left:36.0pt;
	margin-bottom:.0001pt;
	line-height:115%;
	font-size:11.0pt;
	font-family:"Calibri",sans-serif;}
p.MsoListParagraphCxSpLast, li.MsoListParagraphCxSpLast, div.MsoListParagraphCxSpLast
	{margin-top:0cm;
	margin-right:0cm;
	margin-bottom:10.0pt;
	margin-left:36.0pt;
	line-height:115%;
	font-size:11.0pt;
	font-family:"Calibri",sans-serif;}
span.Heading2Char
	{mso-style-name:"Heading 2 Char";
	mso-style-link:"Heading 2";
	font-family:"Calibri Light",sans-serif;
	color:#4472C4;
	font-weight:bold;}
.MsoChpDefault
	{font-family:"Calibri",sans-serif;}
.MsoPapDefault
	{margin-bottom:8.0pt;
	line-height:107%;}
@page WordSection1
	{size:595.3pt 841.9pt;
	margin:72.0pt 72.0pt 72.0pt 72.0pt;}
div.WordSection1
	{page:WordSection1;}
 /* List Definitions */
 ol
	{margin-bottom:0cm;}
ul
	{margin-bottom:0cm;}
-->
</style>

</head>

<body lang=EN-GB>

<div class=WordSection1>

<p class=MsoNormal style='text-align:justify'>DisCVR GUI can be used to carry
out a single sample classification. To launch DisCVR GUI, either double click
on <i>DisCVR.jar</i>, or open a command prompt and type the following commands:</p>

<table class=MsoTableGrid border=1 cellspacing=0 cellpadding=0
 style='border-collapse:collapse;border:none'>
 <tr>
  <td width=308 valign=top style='width:231.05pt;border:none;border-bottom:
  solid windowtext 1.0pt;padding:0cm 5.4pt 0cm 5.4pt'>
  <p class=MsoNormal><span style='font-family:"Courier New"'>cd
  full/path/to/DisCVR folder </span></p>
  <p class=MsoNormal><span style='font-family:"Courier New"'>&nbsp;</span></p>
  <p class=MsoNormal style='page-break-after:avoid'><span style='font-family:
  "Courier New"'>java –jar full/path/to/DisCVR.jar</span></p>
  <p class=MsoNormal style='page-break-after:avoid'>&nbsp;</p>
  </td>
 </tr>
</table>

<p class=MsoCaption><span style='color:windowtext'>Command </span><span style='color:windowtext'>1</span><span style='color:windowtext'>: to move to
DisCVR directory and launch the GUI</span></p>

<p class=MsoNormal style='text-align:justify'>Once DisCVR is launched, there
are 4 panels in the graphical interface: </p>

<p class=MsoListParagraphCxSpFirst style='margin-left:18.0pt;text-align:justify;
text-indent:-18.0pt'>1.<span style='font:7.0pt "Times New Roman"'>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;
</span>Progress Bar: shows the progression of the classification process. </p>

<p class=MsoListParagraphCxSpMiddle style='margin-left:18.0pt;text-align:justify;
text-indent:-18.0pt'>2.<span style='font:7.0pt "Times New Roman"'>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;
</span>Input Sources: allows the user to choose the sample file and a <i>k</i>-mers
database for classification.</p>

<p class=MsoListParagraphCxSpMiddle style='margin-left:18.0pt;text-align:justify;
text-indent:-18.0pt'>3.<span style='font:7.0pt "Times New Roman"'>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;
</span>Classification Results: shows the classification output in three
sub-panels</p>

<p class=MsoListParagraphCxSpLast style='margin-left:18.0pt;text-align:justify;
text-indent:-18.0pt'>4.<span style='font:7.0pt "Times New Roman"'>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;
</span>Progress Information: updates the user about classification output. </p>

<p class=MsoNormal style='text-align:justify'>To start a classification
process, the users need to first select the sample file they wish to
investigate.  Currently, DisCVR supports .fasta, .fa, .fastq, and .fq formats. It
allows for the selection of one file hence, for paired-end files, it is
recommended to concatenate both files. The second step is to choose the database
for the classification. The default setting is to select one of the three
built-in databases. This sets the <i>k</i>-mer size to 22 and removes all
sample <i>k</i>-mers with entropy &#8804; 2.5. On the other hand, if the users
elect to use a customised database for the classification, then the box next to
the <b>Customised Database</b> label must be checked. The fields for this
selection are then activated for the users to upload the customised database file
which MUST be in the <i>customisedDB</i> folder that is on the same path as <i>DisCVR.jar</i>.
The <i>k</i>-mer size for the classification is extracted from the database
name and the threshold to filter out low-entropy <i>k</i>-mers from the sample should
be entered in its corresponding field. The entropy threshold should be the same
value used in the build of the customised database which is recommended to be
in the range of [0, 3.0]. If the users choose not to specify a value for the
entropy threshold then the default value, 2.5, is displayed on the field and
used in the classification. </p>

<p class=MsoNormal style='text-align:justify'> Once, the database library is
selected, the users click the classify button to start the classification
process.  </p>

<p class=MsoNormal style='text-align:justify'>There are 4 stages in the
classification process and the progress bar is updated after the completion of
each stage. Messages in relation to the output of each stage are displayed on
the progress information panel. This includes the number of reads in the sample,
the number of distinct <i>k</i>-mers and their total counts in the database and
the sample file, the number of sample <i>k</i>-mers after removing single
copies and low entropy <i>k</i>-mers, and the number of classified <i>k</i>-mers
and their total counts. The final message shows the number of viruses with classified
<i>k</i>-mers found in the sample and the time taken to finish the
classification process. Figure 1 shows screenshots of the sample classification
process using DisCVR GUI.</p>

<p class=MsoNormal style='page-break-after:avoid'><img width=598 height=576
src="DisCVR_GUI_files/image001.png"></p>

<p class=MsoCaption><span style='color:windowtext'>Figure </span><span style='color:windowtext'>1</span><span style='color:windowtext'>: Screenshots of
a sample classification process using DisCVR GUI. First (a) sample file is uploaded,
<i>k</i>-mers database is then selected either from (b) DisCVR database library
or (c) user’s customised database. Finally, (d) classification results are
displayed.</span></p>

<h2><span style='color:windowtext'>Classification Output</span></h2>

<p class=MsoNormal style='text-align:justify'>At the end of the classification
process, the progress information panel states the number of viruses with
matched <i>k</i>-mers to the database. Detailed information about the classification
results are displayed on the centre panel. </p>

<p class=MsoNormal style='text-align:justify'>The classification results panel
consists of three sub-panels: Scoring, Summary, and Full Analysis. Once the
classification is completed, a bar chart showing up to three viruses is
displayed on the scoring panel. These are the viruses with the most number of classified
distinct <i>k</i>-mers found in the sample. The chart shows for each virus, the
number of distinct <i>k</i>-mers that is specific to the virus as well as the
number of shared <i>k</i>-mers with other viruses in the sample, which are
referred to as non-specific <i>k</i>-mers.   The Summary panel gives a list of the
viruses found in the sample along with their number of specific and
non-specific distinct <i>k</i>-mers. The Full Analysis panel shows a table with
taxonomic and statistical detailed information about the viruses with classified
<i>k</i>-mers in the sample.  The results are uploaded on the table such that
the first row shows the virus with the highest number of <i>k</i>-mers found in
the sample. However, the users can click on any column heading to sort out the
results in the table according to the information in the column.  The full
analysis table consists of 8 columns:</p>

<p class=MsoListParagraphCxSpFirst style='text-align:justify;text-indent:-18.0pt'>1.<span
style='font:7.0pt "Times New Roman"'>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; </span>Virus
Name: the scientific name for the virus taking from the NCBI <i>names.dmp</i>
file.</p>

<p class=MsoListParagraphCxSpMiddle style='text-align:justify;text-indent:-18.0pt'>2.<span
style='font:7.0pt "Times New Roman"'>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; </span>Taxa
ID: the taxonomy identification for the virus in the NCBI taxonomy tree. </p>

<p class=MsoListParagraphCxSpMiddle style='text-align:justify;text-indent:-18.0pt'>3.<span
style='font:7.0pt "Times New Roman"'>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; </span>Virus
Rank: the virus’s rank according to the NCBI <i>nodes.dmp</i> file.</p>

<p class=MsoListParagraphCxSpMiddle style='text-align:justify;text-indent:-18.0pt'>4.<span
style='font:7.0pt "Times New Roman"'>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; </span>Total
counts of <i>k</i>-mers in DB: the total counts of <i>k</i>-mers that represent
the virus in DisCVR’s database.</p>

<p class=MsoListParagraphCxSpMiddle style='text-align:justify;text-indent:-18.0pt'>5.<span
style='font:7.0pt "Times New Roman"'>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; </span>No.
of distinct Classified <i>k</i>-mers: the number of distinct <i>k</i>-mers that
represent the virus in the sample after removing single copies and low entropy <i>k</i>-mers
and matched with the <i>k</i>-mers database.</p>

<p class=MsoListParagraphCxSpMiddle style='text-align:justify;text-indent:-18.0pt'>6.<span
style='font:7.0pt "Times New Roman"'>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; </span>(%)
of distinct Classified <i>k</i>-mers: the percentage of distinct classified <i>k</i>-mers
that represent the virus in the sample’s distinct <i>k</i>-mers. </p>

<p class=MsoListParagraphCxSpMiddle style='text-align:justify;text-indent:-18.0pt'>7.<span
style='font:7.0pt "Times New Roman"'>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; </span>Total
counts of Classified <i>k</i>-mers: the total number of <i>k</i>-mers that
represent the virus in the sample; some distinct <i>k</i>-mers can occur in the
reads multiple times.</p>

<p class=MsoListParagraphCxSpLast style='text-align:justify;text-indent:-18.0pt'>8.<span
style='font:7.0pt "Times New Roman"'>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; </span>(%)
of total classified <i>k</i>-mers: the percentage of the total number of <i>k</i>-mers
that represent the virus in the sample’s total number of <i>k</i>-mers. </p>

<p class=MsoNormal style='text-align:justify'>The table can be saved as .csv
file from the File icon on the GUI tool bar. Figure 2 shows screenshots of the
classification results sub-panels. </p>

<h2><span style='color:windowtext'>Output Validation</span></h2>

<p class=MsoNormal style='text-align:justify'>DisCVR allows the validation of
the classification output using the reference-assembly technique to assess
significance of matches.  During the build of the virus-specific <i>k</i>-mers
database in DisCVR, a library that contains the complete reference genomes of
some of the viruses, which are represented by <i>k</i>-mers in the database, is
generated.  The tool allows two validations using the extracted reference
genome library: </p>

<p class=MsoListParagraphCxSpFirst style='text-align:justify;text-indent:-18.0pt'>1.<span
style='font:7.0pt "Times New Roman"'>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; </span><i>k</i>-mer
Assembly: maps all classified <i>k</i>-mers to a reference genome.</p>

<p class=MsoListParagraphCxSpLast style='text-align:justify;text-indent:-18.0pt'>2.<span
style='font:7.0pt "Times New Roman"'>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; </span>Read
Assembly: maps all sequence reads to a reference genome using Tanoti (a
BLAST-guided aligner).</p>

<p class=MsoNormal style='text-align:justify'>The validation takes place from
the Full Analysis panel when the users right-click on the virus, which they
wish to validate its results. A list with two items: <i>k</i>-mer Assembly and
Read Assembly appears and once the users make a selection, the
reference-assembly starts. When the alignment is finished, a line graph showing
the coverage and depth of the reference genome by the data in the sample is
displayed. A full coverage of the reference genome indicates a true existing of
the virus in the sample.  The graphs can be saved as a PNG file for later
references. Figure 3 shows an example of the validation stage and its outputs. </p>

<p class=MsoNormal style='page-break-after:avoid'><img width=582 height=860
src="DisCVR_GUI_files/image002.png"></p>

<p class=MsoCaption><span style='color:windowtext'>Figure </span><span style='color:windowtext'>2</span><span style='color:windowtext'>: Screenshots
showing sample classification output on the results sub-panels. The (a) scoring
panel shows the three viruses with the highest number of matched k-mers, (b)
the summary panel, and (c) the full analysis panel with detailed information
about the viruses found in the sample.</span></p>

<p class=MsoNormal style='page-break-after:avoid'><img width=578 height=836
src="DisCVR_GUI_files/image003.png"></p>

<p class=MsoCaption><span style='color:windowtext'>Figure </span><span style='color:windowtext'>3</span><span style='color:windowtext'>: Screenshots
showing how to assess the significance of matches from sample classification
results.  (a) right-clicking on a row from the Full Analysis table pops up the
list of choices to start reference assembly. (b) Right-clicking on a graph e.g.
Read-Assembly graph shows the number of reads at a certain reference genome
position. (c) Right-clicking on a graph e.g. <i>k</i>-mer assembly allows the
users to save it as PNG file.</span></p>

</div>

</body>

</html>
